packages/core/src/tools/pageExitState.ts

```diff
@@ -0,0 +1,64 @@
+import type { Subscription } from './observable'
+import { Observable } from './observable'
+import { addEventListener, DOM_EVENT } from './utils'
+
+export interface PageExitState {
+  onPageExit: (callback: (event: PageExitEvent) => void) => Subscription
+  stop: () => void
+}
+
+export interface PageExitEvent {
+  isUnloading: boolean
+}
+
+const enum PageExitStatus {
+  NOT_EXITED,
+  EXITED_HIDDEN,
+  EXITED_UNLOADING,
+}
+
+export function createPageExitState(): PageExitState {
+  let status = PageExitStatus.NOT_EXITED
+  const onPageExitObservable = new Observable<PageExitEvent>()
+
+  function setUnloadStatus(newUnloadStatus: PageExitStatus) {
+    if (status === PageExitStatus.EXITED_UNLOADING || status === newUnloadStatus) {
```

Comment:
I initially thought that it would be fine because we are closing the current view at the first `beforeunload` event so whatever happens after it is kind of a grey area.

But I think that it would be best not to change this behavior, so I changed it.